cibyl/orchestrator.py

```diff
@@ -152,3 +152,25 @@ def extend_parser(self, attributes, group_name='Environment',
                    hasattr(class_type, 'API'):
                     self.extend_parser(class_type.API, class_type.__name__,
                                        level=level+1)
+
+    def register_plugins(self, plugins):
+        """
+        Register plugins.
+        """
+        for plugin in plugins:
+            if hasattr(plugin, 'API_CLASS'):
+                for module, class_name in plugin.API_CLASS.items():
+                    loaded_module = __import__(f"{plugin.__name__}.{module}",
+                                               fromlist=[''])
+                    if hasattr(loaded_module, class_name):
+                        self.extend_parser(
+                            attributes=getattr(loaded_module, class_name).API,
+                            group_name='Plugins',
+                        )
+                    for a_name, a_value in Environment.API.items():
```

Comment:
In order to get to the right model (Job in this case) it needs to traverse the hierarchy of models. Just iterating on the Environment API will not get there (there is System in between). I would move this logic into method of the plugin itself

Reply:
I think the other way around. If this work has to be done by all plugins, then I think it is a good idea to leverage it from them. Otherwise each plugin will have to implement the same thing over and over again.

Reaching jobs should not be that difficult, do we just not need another nested loop that iterates over the systems?

Reply:
True, we should probably give it another thought.

Reply:
@jsanemet I agree with you we need to give it another thought. Currently plugins are gets loaded and it's part of the orchestrator.

Reply:
About this, let me clarify what I meant. The idea of the plugins in cibyl is that they provide product specific information and extend the CI models with it. For the moment, the only plugin that we need is Openstack, that will provide Deployment and other models as part of a Job. Thus, my impression is that a plugin should define the model to add (Deployment) and the model to add it to (Job) and then in `register_plugins` function we can have some generic code that for each plugin traverses the model hierarchy until it finds the right entrypoint and then add to its API the plugin models. If we do that, then when later extend_parser is called it will traverse again the model hierarchy and add all the arguments (both the CI and the plugin ones)